src/filters/detectors/assurance.hpp

```diff
@@ -79,18 +105,43 @@ class AssuranceDetector : public BaseDetector {
       const Tag& _tag,
       const SerenityConfig& _config)
     : BaseDetector(_tag, AssuranceDetectorConfig(_config)),
-      referencePoint(None()),
-      referencePointCounter(0),
-      lastSeverity(0) {}
+      valueBeforeDrop(None()),
+      quorumNum(0) {
+    this->recalculateParams();
+  }
+
+  Result<Detection> _processSample(double_t in);
 
   virtual Result<Detection> processSample(double_t in);
 
+  virtual Try<Nothing> reset();
+
+  /**
+   * Move each base point to next iterator.
+   */
+  void shiftBasePoints();
+
+  /**
+   * Contention Factory.
+   */
+  Detection createContention(double_t severity);
+
+  /**
+   * It is possible to dynamically change detector configuration.
+   */
+  void recalculateParams();
+
   static const constexpr char* NAME = "AssuranceDetector";
+
  protected:
   std::list<double_t> window;
-  Option<double_t> referencePoint;
-  uint64_t referencePointCounter;
-  double_t lastSeverity;
+  std::list<std::list<double_t>::iterator> basePoints;
+
+  // If none then there was no drop.
+  Option<double_t> valueBeforeDrop;
+
+  int32_t dropVotes;
+  int32_t quorumNum;
```

Comment:
What rule you follow to evaluate what type (uint32 vs uint64) choose? Possible length of var?